custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/ThingHandlerSetLabelCheck.java

```diff
@@ -0,0 +1,212 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.tools.analysis.checkstyle;
+
+import static org.openhab.tools.analysis.checkstyle.api.CheckConstants.JAVA_EXTENSION;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Optional;
+import java.util.Queue;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openhab.tools.analysis.checkstyle.api.AbstractStaticCheck;
+
+import com.puppycrawl.tools.checkstyle.TreeWalker;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.FileText;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CheckUtils;
+
+import antlr.RecognitionException;
+import antlr.TokenStreamException;
+
+/**
+ * Checks the code and generates a warning if a class extends BaseThingHandler
+ * and the instance method setLabel of the Thing interface is used.
+ * The label have to be set only by the user.
+ *
+ * @author Tanya Georgieva - Initial Contribution
+ */
+public class ThingHandlerSetLabelCheck extends AbstractStaticCheck {
+
+    private static final String WARNING_MESSAGE = "Do not use the setLabel of the thing in the ThingHandler";
+    private static final String BASE_THING_HANDLER = "BaseThingHandler";
+    private static final String SEARCHED_PACKAGE = "org.eclipse.smarthome.core.thing.binding." + BASE_THING_HANDLER;
+    private static final String SET_LABEL = "setLabel";
+    private static final String THING = "thing";
+    private static final String UNKNOWN = "UNKNOWN";
+
+    private final Log logger = LogFactory.getLog(ThingHandlerSetLabelCheck.class);
+
+    private String fullJavaExtension = "\\." + JAVA_EXTENSION;
+
+    private HashMap<File, DetailAST> files;
+    private HashMap<File, String> derivedAndBaseClasses;
+    private HashMap<File, DetailAST> classesExtendingBaseThingHandler;
+
+    public ThingHandlerSetLabelCheck() {
+        setFileExtensions(JAVA_EXTENSION);
+    }
+
+    @Override
+    public void init() {
+        files = new HashMap<>();
+        derivedAndBaseClasses = new HashMap<>();
+        classesExtendingBaseThingHandler = new HashMap<>();
+    }
+
+    @Override
+    protected void processFiltered(File file, FileText fileText) throws CheckstyleException {
+        FileContents fileContents = new FileContents(fileText);
+        DetailAST rootAST = null;
+
+        try {
+            rootAST = TreeWalker.parse(fileContents);
+        } catch (RecognitionException | TokenStreamException e) {
+            logger.error("Parsing file not successful");
+            e.printStackTrace();
+        }
+        files.put(file, rootAST);
+    }
+
+    @Override
+    public void finishProcessing() {
+        findDerivedAndBaseClasses();
```

Comment:
I think / suppose that extension could easily bypass the boundaries of maven projects and AFAIK this test is executed on maven module level. So if the parent class (B) of our subject (A) which does extend the target class (C) lives in another module it will fail. 
Module 1:
A
Module 2:
B
Module 3 (ESH):
C